lib/web_ui/test/engine/text_editing_test.dart

```diff
@@ -1757,6 +1769,142 @@ Future<void> testMain() async {
+        'text': 'Hello world',
+        'selectionBase': 9,
+        'selectionExtent': 3,
+      });
+      sendFrameworkMessage(codec.encodeMethodCall(setEditingState));
+
+      const MethodCall show = MethodCall('TextInput.show');
+      sendFrameworkMessage(codec.encodeMethodCall(show));
+
+      final DomHTMLInputElement input = textEditing!.strategy.domElement! as
+          DomHTMLInputElement;
+
+      final Object jsInputEvent = js_util.getProperty<Object>(domWindow, 'InputEvent');
+      final List<dynamic> eventArgs = <dynamic>[
+        'beforeinput',
+        js_util.jsify(<String, dynamic>{
+          'inputType': 'delete',
+        }),
+      ];
+      final MockDomInputEvent testEvent = js_util.callConstructor<MockDomInputEvent>(
+        jsInputEvent,
+        eventArgs,
+      );
+      input.dispatchEvent(testEvent);
+
+      EditingState editingState = EditingState(
+        text: 'Helld',
+        baseOffset: 3,
+        extentOffset: 3,
+      );
+      editingState.applyToDomElement(input);
+      input.dispatchEvent(createDomEvent('Event', 'input'));
+
+      expect(spy.messages, hasLength(1));
+      expect(spy.messages[0].channel, 'flutter/textinput');
+      expect(spy.messages[0].methodName, 'TextInputClient.updateEditingStateWithDeltas');
+      expect(
+        spy.messages[0].methodArguments,
+        <dynamic>[
+          123, // Client ID
+          <String, dynamic>{
+            'deltas': <Map<String, dynamic>>[
+              <String, dynamic>{
+                'oldText': 'Hello world',
+                'deltaText': '',
+                'deltaStart': 3,
+                'deltaEnd': 9,
+                'selectionBase': 3,
+                'selectionExtent': 3,
+                'composingBase': -1,
+                'composingExtent': -1
+              }
+            ],
+          }
+        ],
+      );
+      spy.messages.clear();
+
+      hideKeyboard();
+    });
+
+    test('Supports new line at inverted selection', () {
+      final MethodCall setClient = MethodCall(
+          'TextInput.setClient', <dynamic>[123, createFlutterConfig('text', enableDeltaModel: true)]);
+      sendFrameworkMessage(codec.encodeMethodCall(setClient));
+
+      const MethodCall setEditingState =
+      MethodCall('TextInput.setEditingState', <String, dynamic>{
+        'text': 'Hello world',
+        'selectionBase': 9,
+        'selectionExtent': 3,
+      });
+      sendFrameworkMessage(codec.encodeMethodCall(setEditingState));
+
+      const MethodCall show = MethodCall('TextInput.show');
+      sendFrameworkMessage(codec.encodeMethodCall(show));
+
+      final DomHTMLInputElement input = textEditing!.strategy.domElement! as
+          DomHTMLInputElement;
+
+      final Object jsInputEvent = js_util.getProperty<Object>(domWindow, 'InputEvent');
+      final List<dynamic> eventArgs = <dynamic>[
+        'beforeinput',
+        js_util.jsify(<String, dynamic>{
+          'inputType': 'insertLineBreak',
```

Comment:
@mdebbar This test currently fails because there is no `DomInputEvent` defined for `InputEvent` which contains `inputType`. In `handleBeforeInput` we check the `inputType` in the logic. Is there a way to test this?

Reply:
Thanks for the fix! I was trying to avoid adding to the DOM API but if that's the only way to test this I think its okay. `InputEvent` is a basic building block like `KeyboardEvent` and other related events so it would be good to add.

Reply:
Yeah. It seems to me that extending DOM bindings where needed is better than ad-hoc wrapper just for the test.